File: testdata/fixtures/go/CWE-287.go
// CWE-287: Improper Authentication — func Handler(w http.ResponseWriter...) outside authMiddleware
package fixture

import (
    "net/http"
)

// ruleid: go-missing-auth
func adminHandler(w http.ResponseWriter, r *http.Request) {
    w.Write([]byte("secret admin data"))
}

// ruleid: go-missing-auth
func transferHandler(w http.ResponseWriter, r *http.Request) {
    w.Write([]byte("transfer"))
}

// ruleid: go-missing-auth
func sensitiveHandler(w http.ResponseWriter, r *http.Request) {
    w.Write([]byte("sensitive"))
}

// ok: go-missing-auth — contains authMiddleware check
func safeAdminHandler(w http.ResponseWriter, r *http.Request) {
    if !authMiddleware(r) {
        http.Error(w, "unauthorized", http.StatusUnauthorized)
        return
    }
    w.Write([]byte("secret admin data"))
}

// ok: go-missing-auth — RequireAuth check
func safeTransferHandler(w http.ResponseWriter, r *http.Request) {
    if !RequireAuth(r) {
        http.Error(w, "unauthorized", http.StatusUnauthorized)
        return
    }
    w.Write([]byte("transfer"))
}

func authMiddleware(r *http.Request) bool { return true }
func RequireAuth(r *http.Request) bool { return true }

File: testdata/fixtures/javascript/CWE-287.js
// CWE-287: Improper Authentication — Express app.get without authenticate
const express = require('express');
const app = express();
const router = express.Router();

function authenticate(req,res,next){ next(); }
function authMiddleware(req,res,next){ next(); }
function requireAuth(req,res,next){ next(); }

// ruleid: javascript-missing-auth
app.get('/admin', (req, res) => { res.send('secret'); });
// ruleid: javascript-missing-auth
app.post('/transfer', (req, res) => { res.send('done'); });
// ruleid: javascript-missing-auth
router.get('/sensitive', (req, res) => { res.send('data'); });

// ok: javascript-missing-auth
app.get('/admin-safe', authenticate, (req, res) => { res.send('secret'); });
// ok: javascript-missing-auth
router.post('/safe', authMiddleware, (req, res) => { res.send('ok'); });
